grpc/src/main/java/io/stargate/grpc/service/ExceptionHandler.java

```diff
@@ -0,0 +1,270 @@
+import org.apache.cassandra.stargate.exceptions.ReadTimeoutException;
+import org.apache.cassandra.stargate.exceptions.UnavailableException;
+import org.apache.cassandra.stargate.exceptions.UnhandledClientException;
+import org.apache.cassandra.stargate.exceptions.WriteFailureException;
+import org.apache.cassandra.stargate.exceptions.WriteTimeoutException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ExceptionHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(ExceptionHandler.class);
+
+  static final Metadata.Key<QueryOuterClass.Unavailable> UNAVAILABLE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.Unavailable.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.WriteTimeout> WRITE_TIMEOUT_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.WriteTimeout.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.ReadTimeout> READ_TIMEOUT_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.ReadTimeout.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.ReadFailure> READ_FAILURE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.ReadFailure.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.FunctionFailure> FUNCTION_FAILURE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.FunctionFailure.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.WriteFailure> WRITE_FAILURE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.WriteFailure.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.AlreadyExists> ALREADY_EXISTS_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.AlreadyExists.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.CasWriteUnknown> CAS_WRITE_UNKNOWN_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.CasWriteUnknown.getDefaultInstance());
+
+  private final StreamObserver<QueryOuterClass.Response> responseObserver;
+  private final AtomicBoolean exceptionOccurred = new AtomicBoolean(false);
+
+  public ExceptionHandler(StreamObserver<QueryOuterClass.Response> responseObserver) {
+
+    this.responseObserver = responseObserver;
+  }
+
+  protected void handleException(Throwable throwable) {
+    exceptionOccurred.set(true);
+    if (throwable instanceof CompletionException
+        || throwable instanceof MessageHandler.ExceptionWithIdempotencyInfo) {
+      handleException(throwable.getCause());
+    } else if (throwable instanceof StatusException
+        || throwable instanceof StatusRuntimeException) {
+      synchronized (responseObserver) {
+        responseObserver.onError(throwable);
+      }
+    } else if (throwable instanceof UnhandledClientException) {
+      onError(Status.UNAVAILABLE, throwable);
+    } else if (throwable instanceof PersistenceException) {
+      handlePersistenceException((PersistenceException) throwable);
+    } else {
+      LOG.error("Unhandled error returning UNKNOWN to the client", throwable);
+      synchronized (responseObserver) {
```

Comment:
can this and synchronized on the like 60 be somehow also moved to the private method? would be nice to have synchronized in a single place only..